VSTSDropProxy.cs

```diff
@@ -83,141 +85,106 @@ public VSTSDropProxy(string VSTSDropUri, string path, string pat, TimeSpan blobt
-         // 1) only grab certain directories either with dockerfiles or as specified by build.yaml
-         // 2) Prioritize large files or files with lots of copies.
-         // 3) linux symlink instead of copy
-         // 4) parallelize copy with buffer first attempt at that with _contentClient.GetBufferAsync failed. Also lots of memory.
-         // 5) multistream copyasync
-         // Tried configureawait(false) and copyaync to each file (though not aparallelized) with no effect.
-        public async Task<Dictionary<string,double>> Materialize(string localDestiantion)
+
+        public async Task<Dictionary<string, double>> Materialize(string localDestination)
         {
-            var uniqueblobs = _files.GroupBy(file => file.Blob.Id).ToList();
+            var uniqueblobs = _files.GroupBy(keySelector: file => file.Blob.Id, resultSelector: (key, file) => file).ToList();
             Console.WriteLine($"Found {_files.Count} files, {uniqueblobs.Count} unique");
-            var metrics = new Dictionary<string,double>();
-            metrics["files"] = _files.Count;
-            metrics["uniqueblobs"] = uniqueblobs.Count;
-            var dockerdirs = new HashSet<string>();
-            //precreate directories so we don't have to worry.
-            //slow but we don't care cause we want to move off and have sangam use blobs directly
-            foreach (var file in _files) //could do blobs.selectmany(vallues)
+            var metrics = new Dictionary<string, double>
             {
-                var replativepath = file.Path.Substring(_relativeroot.Length);
-                var localpath = Path.Combine(localDestiantion,replativepath).Replace("\\","/");
-                //also not efficient to check directory each time 
-                
-                Directory.CreateDirectory(Path.GetDirectoryName(localpath));
-                var filename = Path.GetFileName(localpath);
-                if (filename.StartsWith("dockerfile", StringComparison.OrdinalIgnoreCase))
-                {
-                    dockerdirs.Add(Path.GetDirectoryName(file.Path));
-                }
-            }
+                ["files"] = _files.Count,
+                ["uniqueblobs"] = uniqueblobs.Count
+            };
 
-            metrics["dockerfiles"] = dockerdirs.Count;
+            var dltimes = new ConcurrentBag<double>();
+            var copytimes = new ConcurrentBag<double>();
+            var throttler = new ActionBlock<IEnumerable<VstsFile>>(list => DownloadGrouping(list, localDestination, dltimes, copytimes), new ExecutionDataflowBlockOptions { MaxDegreeOfParallelism = ConcurrentDownloadCount });
 
-            //Altenatively would be neat to hash as we iterate throgh first loop
-            foreach (var ddir in dockerdirs)
+            foreach (var grouping in uniqueblobs)
             {
-                var dirfiles = _files.Where(f => f.Path.StartsWith(ddir)).ToList();
-                var hash = HashFiles(dirfiles);
```

Comment:
lost this functionality which is important for us (lets us know if we need to call docker build on a dir)>

Reply:
whoops let me add that back when i get a chance